src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -533,9 +532,50 @@ public async Task WebApplicationBuilder_StartupFilterCanAddTerminalMiddleware()
             Assert.Equal(418, (int)terminalResult.StatusCode);
         }
 
+        [Fact]
+        public async Task WebApplicationBuilder_OnlyAddsDefaultServicesOnce()
+        {
+            var builder = WebApplication.CreateBuilder();
+
+            // IWebHostEnvironment is added by ConfigureDefaults
+            Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IConfigureOptions<LoggerFactoryOptions>)));
+            // IWebHostEnvironment is added by ConfigureWebHostDefaults
+            Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IWebHostEnvironment)));
+            Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IOptionsChangeTokenSource<HostFilteringOptions>)));
+            Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IServer)));
+
+            await using var app = builder.Build();
+
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IConfigureOptions<LoggerFactoryOptions>>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IWebHostEnvironment>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IOptionsChangeTokenSource<HostFilteringOptions>>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IServer>>());
+        }
+
+        [Fact]
+        public void WebApplicationBuilder_EnablesServiceScopeValidationByDefaultInDevelopment()
+        {
+            // The environment cannot be reconfigured after the builder is created currently.
```

Comment:
Not part of this PR but we should talk about ways to throw if `UseEnvironment` is used after this call (and maybe we need a first class API).